src/ellswift.rs

```diff
@@ -0,0 +1,686 @@
+//! - The paper uses an additional encoding bit for the parity of y. Here the
+//!   parity of t is used (negating t does not affect the decoded x coordinate,
+//!   so this is possible).
+
+use core::fmt::{self, Display, Formatter};
+use core::ptr;
+use core::str::FromStr;
+
+use ffi::secp256k1_ellswift_xdh;
+use secp256k1_sys::types::{c_int, c_uchar, c_void};
+
+use crate::ffi::{
+    self, secp256k1_ellswift_create, secp256k1_ellswift_decode, secp256k1_ellswift_encode, CPtr,
+};
+use crate::{constants, from_hex, Context, Error, PublicKey, Secp256k1, SecretKey};
+
+unsafe extern "C" fn hash_callback<F>(
+    output: *mut c_uchar,
+    x32: *const c_uchar,
+    ell_a64: *const c_uchar,
+    ell_b64: *const c_uchar,
+    hash_func: *mut c_void,
+) -> c_int
+where
+    F: FnMut([u8; 32], [u8; 64], [u8; 64]) -> ElligatorSwiftSharedSecret,
+{
+    let callback: &mut F = &mut *(hash_func as *mut F);
+    let mut x32_array = [0u8; 32];
+    let mut ell_a64_array = [0u8; 64];
+    let mut ell_b64_array = [0u8; 64];
+
+    // Copy the data into Rust slices
+    ptr::copy_nonoverlapping(x32, x32_array.as_mut_c_ptr(), 32);
+    ptr::copy_nonoverlapping(ell_a64, ell_a64_array.as_mut_c_ptr(), 64);
+    ptr::copy_nonoverlapping(ell_b64, ell_b64_array.as_mut_c_ptr(), 64);
+    // Call the hash function that was passed in through the `data` pointer
+    let secret = callback(x32_array, ell_a64_array, ell_b64_array);
+    // Copy the output from a [ElligatorSwiftSharedSecret] into the output pointer
+    ptr::copy_nonoverlapping(secret.0.as_ptr(), output, secret.0.len());
+
+    secret.0.len() as c_int
+}
+/// ElligatorSwift is an encoding of a uniformly chosen point on the curve
+/// as a 64-byte array that is indistinguishable from a uniformly random.
+/// This object holds tow field elements u and t, which are the inputs to
+/// the ElligatorSwift encoding function.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ElligatorSwift(ffi::ElligatorSwift);
+
+impl ElligatorSwift {
+    /// Create a new ElligatorSwift object from a 64-byte array.
+    pub fn ellswift_create(priv32: SecretKey, rand32: [u8; 32]) -> ElligatorSwift {
+        let mut ell_out = [0u8; constants::ELLSWIFT_ENCODING_SIZE];
+        unsafe {
+            secp256k1_ellswift_create(
+                ffi::secp256k1_context_no_precomp,
+                ell_out.as_mut_c_ptr(),
+                priv32.as_c_ptr(),
+                rand32.as_ptr(),
+            );
+        }
+        ElligatorSwift(ffi::ElligatorSwift::from_array(ell_out))
+    }
+    /// Decode an ElligatorSwift object from a 64-byte array.
+    pub fn from_array(ser: [u8; 64]) -> ElligatorSwift {
+        ElligatorSwift(ffi::ElligatorSwift::from_array(ser))
+    }
+    /// Encode an ElligatorSwift object as a 64-byte array.
+    pub fn to_array(&self) -> [u8; 64] { self.0.into_array() }
+    /// Creates the Elligator Swift encoding from a secret key, using some aux_rand if defined.
```

Comment:
nit: This whole file has different white spacing to the rest of the codebase.
```suggestion

    /// Creates the Elligator Swift encoding from a secret key, using some aux_rand if defined.
```

Reply:
Do you mean an empty line between items?

Reply:
Yes please, that diff suggestion I did was a bit average wasn't it. Its only a nit though, you don't have to follow my suggestions :)